capstone/src/main/java/com/google/sps/servlets/TagsTFIDFServlet.java

```diff
@@ -62,39 +65,82 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
     for (long groupId : groupIds) {
       Entity groupEntity = ServletHelper.getEntityFromId(response, groupId, datastore, "Group");
-      List<Long> postIds =
-          (groupEntity.getProperty("posts") == null)
-              ? new ArrayList<Long>()
-              : (ArrayList<Long>) groupEntity.getProperty("posts");
-      List<String> postTexts = getGroupPostsText(postIds, response);
 
-      // TODO: Parallelize this process to occur for posts concurrently.
+      List<String> textData = getGroupPostsText(groupEntity, response);
+      addChallengeText(groupEntity, textData, response);
+
+      // TODO: Parallelize this process to occur for each string concurrently.
```

Comment:
TECHNICALLY you probably want to batch them somehow, whether you just divide the array by 4 and assign the strings to up to 4 threads, or define some max strings per thread and dynamically determine threads based on number of strings.

Making a thread is very expensive and slow, so you need to find the right tradeoff between making thread and not making thread.

Currently, your test dataset isn't large enough to actually justify threading as a speed optimization, but it's still good practice to see if you could define the problem in the event that there's a million users

Reply:
(And if you end up with spare time, I'd rather you implement a threading solution than leave it as a TODO :) )